2.0: Full Rewrite

KillerOfPie opened the pull request:
This PR contains major changes, improvements and fixes. See the [2.0 Release Notes](https://github.com/SparklingComet/TradeShop/wiki/2.0-Release-notes) for more information.

KillerOfPie committed:
Working on 2.0 Update (In-Progress)
- Added Permissions Enum
- Added JsonConfiguration Object
- Updated ShopType Enum to be consistent with other Enums
- Updated all getPerm calls to new Enum

KillerOfPie committed:
Road to 2.0
- Added Permissions Enum
- Added Setting Enum
- Added Custom Item Class
- Added JsonConfiguration

- Updated ShopType Enum
- Cleaned out unnecessary from TradeShop

- Overall code cleanup

KillerOfPie committed:
Merge remote-tracking branch 'origin/2.0' into 2.0

# Conflicts:
#	pom.xml
#	src/main/java/org/shanerx/tradeshop/itrade/IShopCreateEventListener.java
#	src/main/java/org/shanerx/tradeshop/util/Utils.java
#	src/main/resources/plugin.yml

KillerOfPie committed:
Road to 2.0
- New files missing in last commit

SparklingComet commented on src/main/java/org/shanerx/tradeshop/enums/Permissions.java:
```diff
@@ -0,0 +1,57 @@
+    }
+
+    public Permission getPerm() {
+        return new Permission("tradeshop." + key);
```
Maybe use `new Permission(toString())`? (DRY)

KillerOfPie replied:
Done

SparklingComet commented on src/main/java/org/shanerx/tradeshop/enums/Permissions.java:
```diff
@@ -0,0 +1,57 @@
+    WHO("who");
+
+    private String key;
+    private static TradeShop plugin = (TradeShop) Bukkit.getPluginManager().getPlugin("TradeShop");
```
This is a perfect example for a `final` field

SparklingComet replied:
Although this is a private field, final/effectively final variables allow the compiler to apply optimisations, thus it should be used here.

KillerOfPie replied:
Didnt realize it wasnt final, is now

SparklingComet commented on src/main/java/org/shanerx/tradeshop/enums/Setting.java:
```diff
@@ -0,0 +1,153 @@
+        return name().toLowerCase().replace("_", "-");
+    }
+
+    public Object getSetting() {
```
I don’t like the `Object` return type... on the other hand, `String` is unfit as well... how about a “custom” `ConfigurationValue` object?
`String getString()`
`boolean getBoolean()`
`int getInt()`
`double getDouble()`?

And setters as well
So it basically does the casting internally, assuming you call getString when it’s a string, getBoolean when it is a boolean, etc...
It would serve the DRY principle as well as cleaning up the code because there wouldn’t be any casting.

Edit: noticing just now you are doing it just a couple lines below :S 
I’d still prefer a separate object, which can be retrieved with Setting#getValue

KillerOfPie replied:
The object one is not there to be used, it can actually probably be removed and i cant to a custom type because it varies on the field, which is why theres specific get methods for each type. also setting is always done with an object, no reason to make a method without object because its going to do the exact same thing.

SparklingComet replied:
I am aware of that; what I am suggesting is a wrapper object.

KillerOfPie replied:
I'm already wrapping all the gets, what would be the point of wrapping a set?

SparklingComet replied:
It would be cleaner and would look better, which is partially the point of this rewrite. No difference in the semantics, though.

KillerOfPie replied:
But all the methods would do the same exact thing so would it be more efficient to just have the one set method with object?

SparklingComet commented on src/main/java/org/shanerx/tradeshop/enums/Setting.java:
```diff
@@ -0,0 +1,153 @@
+    }
+
+    public Object getSetting() {
+        return config.get(name().toLowerCase().replace("_", "-"));
```
Code is screaming “DRY” :P
Replace with #toStrinf(), please

KillerOfPie replied:
Done, these are all the result of 1am coding

SparklingComet commented on src/main/java/org/shanerx/tradeshop/enums/Setting.java:
```diff
@@ -0,0 +1,153 @@
+    }
+
+    public String getString() {
+        return config.getString(name().toLowerCase().replace("_", "-"));
```
Same here

KillerOfPie replied:
done

SparklingComet commented on src/main/java/org/shanerx/tradeshop/enums/Setting.java:
```diff
@@ -0,0 +1,153 @@
+    }
+
+    public List<String> getStringList() {
+        return config.getStringList(name().toLowerCase().replace("_", "-"));
```
And here...

KillerOfPie replied:
done

SparklingComet commented on src/main/java/org/shanerx/tradeshop/enums/Setting.java:
```diff
@@ -0,0 +1,153 @@
+    }
+
+    public int getInt() {
+        return config.getInt(name().toLowerCase().replace("_", "-"));
```
^

KillerOfPie replied:
done

SparklingComet commented on src/main/java/org/shanerx/tradeshop/enums/Setting.java:
```diff
@@ -0,0 +1,153 @@
+    }
+
+    public double getDouble() {
+        return config.getDouble(name().toLowerCase().replace("_", "-"));
```
^

KillerOfPie replied:
done

SparklingComet commented on src/main/java/org/shanerx/tradeshop/enums/Setting.java:
```diff
@@ -0,0 +1,153 @@
+    }
+
+    public boolean getBoolean() {
+        return config.getBoolean(name().toLowerCase().replace("_", "-"));
```
^

KillerOfPie replied:
done

SparklingComet commented on src/main/java/org/shanerx/tradeshop/enums/Setting.java:
```diff
@@ -0,0 +1,153 @@
+        return config.getBoolean(name().toLowerCase().replace("_", "-"));
+    }
+
+    public static ArrayList<String> getIllegalItems() {
```
Not entirely sure whether this is convenient - if we wrap up `Setting.ILLEGAL_ITEMS.getStringList()`, maybe in the `TradeShop` plugin) class it would make more sense because of `plugin.getItemBlacklist()` (sounds/looks better as well, imo)

KillerOfPie replied:
I can change it to Blacklist, just copied and pasted it from TradeShop, and the goal of moving it was getting unneccesary things out of the TradeShop file

KillerOfPie replied:
done

SparklingComet commented on src/main/java/org/shanerx/tradeshop/enums/Setting.java:
```diff
@@ -0,0 +1,153 @@
+    private static File file = new File(plugin.getDataFolder(), "config.yml");
+    private static FileConfiguration config = YamlConfiguration.loadConfiguration(file);
+
+    private static void setDefaults() {
```
I know this is better in terms of performance, but we *could* consider using `FileConfiguration#saveDefaultConfig()`in the *future*... mostly because it would introduce comments support.

(**Not now**, in the future)

KillerOfPie replied:
I will clean this up later but i do prefer this over #saveDefaultConfig()

SparklingComet replied:
So do I, although comments are handy.
As I said, we will not change this in 2.0, but se’ll keep the idea in the back of our head.

KillerOfPie replied:
kk

SparklingComet commented on src/main/java/org/shanerx/tradeshop/enums/Setting.java:
```diff
@@ -0,0 +1,153 @@
+        }
+
+        config = YamlConfiguration.loadConfiguration(file);
+        setDefaults();
```
The rest looks good, +1

SparklingComet commented on src/main/java/org/shanerx/tradeshop/enums/ShopType.java:
```diff
@@ -19,55 +19,55 @@
-        return getPlugin().getSettings().getString("tradeshop-name");
+    @Override
+    public String toString() {
+        return "[" + Setting.findSetting(key + "shop-name").getString() + "]";
```
Looking at DRY here - definitely cleaner.

SparklingComet replied:
Re-use `stripped()`here

KillerOfPie replied:
one day ill stop skipping over that when remaking stuff

SparklingComet replied:
What do you mean by “remaking stuff”?

KillerOfPie replied:
IDK it was early in the morning/late at night

SparklingComet commented on src/main/java/org/shanerx/tradeshop/itrade/IShopCreateEventListener.java:
```diff
@@ -47,24 +49,24 @@ public IShopCreateEventListener(TradeShop instance) {
     @SuppressWarnings("deprecation")
     @EventHandler
     public void onSignChange(SignChangeEvent event) {
-        String header = ShopType.ITRADE.header();
```
I was thinking of merging the create listeners... it would make the classes longer and potentially more difficult to maintain, but it would improve performance due to Bukkit’s use of reflection.

KillerOfPie replied:
once i get to remaking this stuff creation will probably be merged but in a much cleaner way

SparklingComet replied:
Sounds good to me.

SparklingComet commented on src/main/java/org/shanerx/tradeshop/itrade/ITradeEventListener.java:
```diff
@@ -30,9 +30,10 @@
-import org.shanerx.tradeshop.Utils;
+import org.shanerx.tradeshop.enums.Message;
+import org.shanerx.tradeshop.enums.Setting;
+import org.shanerx.tradeshop.util.Utils;
```
We’ll need to split `Utils`and implement OOP, so let’s make sure to keep that among the first items in the todo list.

KillerOfPie replied:
lots of utils will be stripped down when i get to actually remaking the trade and chest stuff and will definitely be more object oriented

SparklingComet commented on src/main/java/org/shanerx/tradeshop/util/CustomItem.java:
```diff
@@ -0,0 +1,142 @@
+        }
+    }
+
+    public static void removeItem(String name) {
```
Nonononono, static-abuse!!! Sorry, but we need to have a `CustomItem` object in order to invoke `CustomItem#remove()` on it!

KillerOfPie replied:
This was just getting things out of TradeShop for now Ill work on actually redoing to CustomItem system later

SparklingComet commented on src/main/java/org/shanerx/tradeshop/util/CustomItem.java:
```diff
@@ -0,0 +1,142 @@
+        }
+    }
+
+    public static ItemStack getItem(String name) {
```
Static abuse again

SparklingComet commented on src/main/java/org/shanerx/tradeshop/util/CustomItem.java:
```diff
@@ -0,0 +1,142 @@
+        }
+    }
+
+    public static Set<String> getItemSet() {
```
And again static abuse... I think we need a `CustomItemManager` class which manages custom items.
`CustomItemManager#getItemSet()` retrieves the CustomItem object.

I suggested `CustomItem#remove()`, but `CustomItemManager#removeItem(CustomItem)` works too.

SparklingComet commented on src/main/java/org/shanerx/tradeshop/util/CustomItem.java:
```diff
@@ -0,0 +1,142 @@
+        return config.getValues(false).keySet();
+    }
+
+    private static void setDefaults() {
```
Please refer to previous comment

SparklingComet commented on src/main/java/org/shanerx/tradeshop/util/CustomItem.java:
```diff
@@ -0,0 +1,142 @@
+        return itm;
+    }
+
+    public static void reload() {
```
Same here, add this as a non-static method to the CustomItemManager class... it would improve code quality and plugin architecture/algorithms.

SparklingComet commented on src/main/java/org/shanerx/tradeshop/util/CustomItem.java:
```diff
@@ -0,0 +1,142 @@
+        setDefaults();
+    }
+
+    private static void save() {
```
^

SparklingComet commented on src/main/java/org/shanerx/tradeshop/util/CustomItem.java:
```diff
@@ -0,0 +1,142 @@
+        config = YamlConfiguration.loadConfiguration(file);
+    }
+
+    public static FileConfiguration getConfig() {
```
(Static abuse)

SparklingComet commented on src/main/java/org/shanerx/tradeshop/util/JsonConfiguration.java:
```diff
@@ -0,0 +1,96 @@
+import java.io.FileReader;
+import java.io.FileWriter;
+
+public class JsonConfiguration
```
Now that I see this, I remember an old project I wrote - that does exactly this... silly me that I didn’t remember this. It has a couple utility methods we could copy over.
([Click](https://github.com/SparklingComet/ConfigAPI)

SparklingComet replied:
This class should implement the `Serializable` Java standard API interface.

KillerOfPie replied:
This is straight copy paste tweakery from stuff online, ive never dealt with json so haveing something to properly reference will make this much easier, also have no clue how ell any of this does work

SparklingComet commented on src/main/java/org/shanerx/tradeshop/util/JsonConfiguration.java:
```diff
@@ -0,0 +1,96 @@
+        return var;
+    }
+
+    public String convertLocation(Location loc) {
```
Is “serializeLocation” the name you are looking for?

SparklingComet replied:
Serialization means to convert an object to a strinf in order to store it on a hard disk, which is exactly the purpose here. The point is, that you can, later on, deserialize it, aka restore it from that string.

KillerOfPie replied:
yes, aaannnnd changed

SparklingComet commented on src/main/java/org/shanerx/tradeshop/util/JsonConfiguration.java:
```diff
@@ -0,0 +1,96 @@
+        return world + div + x + div + y + div + z;
+    }
+
+    public Location convertLocation(String loc) {
```
Same here for “deserializeLocation”

KillerOfPie replied:
done

SparklingComet commented on src/main/java/org/shanerx/tradeshop/util/Updater.java:
```diff
@@ -1,19 +1,24 @@
 /*
- *     Copyright 2016-2017 SparklingComet @ http://shanerx.org
+ *     Copyright (c) 2016-2017 SparklingComet @ http://shanerx.org
```
Theoretically, (c) = © = Copyright, so we are repeating the same word twice.

On a serious note, though, I thought you were part of this copyright - I remember clearly adding you to the header.

Make sure to add yourself before the 2.0 release ^_^

KillerOfPie replied:
well see if i remember and also i didnt put that c there sooooo ya...

SparklingComet replied:
Oh? Well, that’s unexpected... Not sure why’d appear there, then.

SparklingComet commented on src/main/java/org/shanerx/tradeshop/util/Utils.java:
```diff
@@ -118,60 +116,6 @@ public String getPrefix() {
-    public Permission getAdminPerm() {
-        return PADMIN;
-    }
-
```
Uh... why dozens of lines of code being obliterated?
I assume because it is part of the Permissions enum... (?)

KillerOfPie replied:
correct

SparklingComet commented on src/main/java/org/shanerx/tradeshop/Message.java:
```diff
@@ -1,60 +0,0 @@
-    public static String colour(String x) {
-        return ChatColor.translateAlternateColorCodes(COLOUR_CHAR, x);
-    }
-}
```
Where is all of this going? Is it being rewritten, or...?

KillerOfPie replied:
It moved to the Enum package

SparklingComet commented on src/main/java/org/shanerx/tradeshop/TradeShop.java:
```diff
@@ -152,7 +113,7 @@ private void addMaterials() {
         allowedOld.addAll(Arrays.asList(Material.CHEST, Material.TRAPPED_CHEST, Material.DROPPER, Material.HOPPER, Material.DISPENSER));
 
-        for (String str : getSettings().getStringList("allowed-shops")) {
+        for (String str : Setting.ALLOWED_SHOPS.getStringList()) {
```
Can we use a lambda for this?

SparklingComet commented on src/main/java/org/shanerx/tradeshop/TradeShop.java:
```diff
@@ -183,205 +144,9 @@ private void addDirections() {
         allowed.addAll(Arrays.asList(BlockFace.DOWN, BlockFace.WEST, BlockFace.SOUTH, BlockFace.EAST, BlockFace.NORTH, BlockFace.UP));
 
-        for (String str : getSettings().getStringList("allowed-directions")) {
+        for (String str : Setting.ALLOWED_DIRECTIONS.getStringList()) {
```
Same here, a java-8 one-liner is definitely fancier than the old-fashioned “enhanced” for-loop.

SparklingComet replied:
The lambda expression doesn’t have to be one line long, multi-line lambdas are absolutely fine.

SparklingComet commented on src/main/java/org/shanerx/tradeshop/bitrade/BiTradeEventListener.java:
```diff
@@ -60,12 +61,12 @@ public void onBlockInteract(PlayerInteractEvent e) {
                 chestState = findShopChest(s.getBlock()).getState();
             } catch (NullPointerException npe) {
-                buyer.sendMessage(colorize(getPrefix() + plugin.getMessages().getString("missing-shop")));
+                buyer.sendMessage(colorize(getPrefix() + Message.MISSING_SHOP.toString()));
```
how about making Message.toString() prefix the string automatically? More efficient than using `getPrefix() + Message` every time (DRY again?)

SparklingComet commented on src/main/java/org/shanerx/tradeshop/enums/Message.java:
```diff
@@ -0,0 +1,171 @@
+    WHO_MESSAGE;
+    
+    @Override
+    public String toString() {
```
Refer to the comment above.

SparklingComet reviewed (dismissed):
I have taken some time to review the code quality of the PR. I am quite pleased with the current progress and am surprised that you are rewriting everything on your own without any problems, so +1 on that. (It’s not that I wasn’t trusting your skills).

Reading some (most) of my comments may give you the impression that I am obsessed with the “DRY principle” and “OOP” practices. This is partly true, because, in my opinion, the lack of their enforcement is what almost forced us to this rewrite - they will make the code much cleaner, smoother, more efficient and easy to maintain/debug so it’s definitely worth the time spent to implement them.

Once again, thank you so much for your fundamental contributions.

KillerOfPie committed:
Road to 2.0
- Removed CustomItem file
- Added CustomItemManager Object
- Added ListManager Object
- Added ShopRole Enum
- Added ShopStatus Enum
- Added User Object

- Fixes and tweaks mostly related to changing/adding new files

KillerOfPie committed:
Road to 2.0
- Added #getPrefixed to messages
- updated(hopefully) all references to Message#toString
- removed a lot of unnecessary #colorize/#colour

SparklingComet commented on src/main/java/org/shanerx/tradeshop/object/JsonConfiguration.java:
```diff
@@ -0,0 +1,97 @@
+import java.io.FileWriter;
+
+public class JsonConfiguration
+        extends Utils {
```
Remember to add `implements Serializable`

SparklingComet commented:
> fixed, its now isBelow19 and checks if the version is 1.9, 1.10, 1.11, or 1.12. just means we have to add new versions as they come out

How about actually parsing the version number as a `double` and checking whether it’s greater than or equal to `1.9`?

SparklingComet commented:
> should i change the red to &c as well?

Would make sense in terms of consistency, but I don’t really mind... I just want the shade of green changed in order to make the sign coloured.

SparklingComet commented:
>!No there will be an object for the whole shop that holds this, which will be were most of the shop stuff will be

I see, makes sense. So basically, the `User` object is simply a data storage utility. I like the idea, though - definitely OO.
Maybe we could rename it to `ShopUser` as `User` kind of plain and used in other APIs.

KillerOfPie committed:
Road to 2.0
- Simplified methods
- Added Shop File
- Some other stuff that I don't remember

KillerOfPie committed:
Road to 2.0
- Added ShopChunk Object
- Added ShopLocation Object

SparklingComet committed:
Re-implement serialization by switching to Gson
+ fix variable bug
+ reformat code to use tabs instead of spaces in ALL source files

KillerOfPie commented:
FYI all of this is untested

KillerOfPie committed:
Road to 2.0
- Added BukkitVersion object
- Took old version checking out to TradeShop file
- Changed User file to ShopUser

KillerOfPie committed:
Road to 2.0
- Moved Files

SparklingComet commented on src/main/java/org/shanerx/tradeshop/object/BukkitVersion.java:
```diff
@@ -0,0 +1,82 @@
+        return minor;
+    }
+
+    public Boolean isBelow19() {
```
how about we implement a method with the following signature:
`public boolean isAbove(double version);` ?

SparklingComet replied:
Also why are we using the boxed primitives? (`Boolean` instead of `boolean`)

SparklingComet reviewed (commented):
Indentation needs fixing, I assume your IDE uses spaces instead of tabs. If possible, please use tabs instead, as spaces are not compliant with the Java Coding Conventions.

KillerOfPie committed:
Road to 2.0
- Added methods for getting a shops/shop inventories location as ShopLocation Objects
- Removed (I think) static modifiers that shouldnt have been there
- Changed back to old ShopChunk/ShopLocation serialize and deserialize methods
- Added getLocation method to ShopLocation
- Added isAbove method to BukkitVersion
- Hopefully got JsonConfiguration to a point were it should work as we need it to

SparklingComet commented on src/main/java/org/shanerx/tradeshop/object/ShopLocation.java:
```diff
@@ -21,47 +21,56 @@
 		return world;
 	}
 
+	public String getWorldString() {
```
maybe Change to #getWorldName

KillerOfPie commented:
@SparklingComet agreed, I'll change that tomorrow anything else though?

KillerOfPie commented:
@SparklingComet agreed, I'll change that tomorrow anything else though?

SparklingComet commented on src/main/java/org/shanerx/tradeshop/util/BukkitVersion.java:
```diff
@@ -56,10 +56,14 @@ public int getMinor() {
     }
 
-    public Boolean isBelow19() {
+    public boolean isBelow19() {
```
isn't this method redundant? `!isAbove(1.9)` is a valid replacement, no need for duplicates

KillerOfPie replied:
I'll fix it

SparklingComet commented on src/main/java/org/shanerx/tradeshop/util/JsonConfiguration.java:
```diff
@@ -56,55 +60,40 @@ public JsonConfiguration(Chunk c) {
-		String world = chunk.getWorld().getName();
-		int x = chunk.getX(), z = chunk.getZ();
+	private void setShop(Shop shop) {
+		Gson gson = new GsonBuilder().setPrettyPrinting().serializeNulls().create();
```
Not sure why pretty printing is necessary - we use JSON for persistent data storage functions, not configurations/user Interfaces, which means that, if the file size increases, performance will be affected because of all the whitespace and other chars. (It is still micro-performance, though, hence it is a matter of taste.)

KillerOfPie replied:
@SparklingComet didn't think about that, I still like it to be a bit more readable but it isn't necessary so I'll take it out

SparklingComet replied:
Meh
It doesn’t matter much
If you prefer pretty printing, I don’t have any objections.

SparklingComet commented on src/main/java/org/shanerx/tradeshop/util/JsonConfiguration.java:
```diff
@@ -56,55 +60,40 @@ public JsonConfiguration(Chunk c) {
 
-		return "c" + div + world + div + x + div + z;
-	}
+		jsonObj.add(shop.getShopLocationAsSL().serialize(), gson.toJsonTree(shop));
```
uh... what does SL mean?

KillerOfPie replied:
@SparklingComet ShopLocation didn't want a 2 mile long method also wierd to read shop location as shop location

SparklingComet replied:
Call it #getShopLocation ?

KillerOfPie replied:
#getShopLocation returns a Location object because were going to serialize Shop with gson and id rather that get serialized so that ShopLocation doesnt need to be tweaked to work well with the gson serialization.

KillerOfPie committed:
Road to 2.0
- Small fixes from last commit

SparklingComet committed:
Split up Utils.java+ more changes which do not affect the semantics.

SparklingComet committed:
Enlarge disabled metrics message

SparklingComet committed:
Merge all shop creation listeners (SignChangeEvent)

KillerOfPie committed:
- Fixed ambiguous code for getTargetBlock, Casted null

SparklingComet committed:
Add files missed in last commit

SparklingComet committed:
Fix merge conflicts

SparklingComet committed:
Fix merge conflicts

KillerOfPie committed:
Fixed Broken POM

KillerOfPie committed:
Fixed ShopType

KillerOfPie committed:
Fixed TradeShop, Removed Extra Message Enum

KillerOfPie committed:
Fixed Setting, added explode settings

KillerOfPie committed:
Setting - Replaced old toString with toPath
Message fixed old Setting

KillerOfPie committed:
First Build - Untested
Fixed old Settings

KillerOfPie committed:
Started merging listeners - Unfinished

KillerOfPie committed:
Merged i/bi listener into trade listener

KillerOfPie committed:
Fixed random commit stuff in plugin.yml

KillerOfPie committed:
Packages renamed, excess files removed, and small fixes. first load.

KillerOfPie committed:
Lots of fixes to allow shop creation, Tested to creation of all 3 shops

KillerOfPie committed:
Changes to make shop creation and trading work

KillerOfPie committed:
Changing over to new shop and storage system

KillerOfPie committed:
Sign creation works, as well as data saving

KillerOfPie committed:
Sign creation and adding product/cost all implemented and test(still some more features to add back to all of them, bunch of bug fixes along the way

SparklingComet committed:
Fix typo in package name

SparklingComet committed:
Make ShopTradeListener use the Shop API

KillerOfPie committed:
Working on updating all the old files and methods for the new systems, working trading from file

KillerOfPie committed:
adding support for open/closed shops and other stuff

KillerOfPie committed:
Blacklist is working, added open and close commands, shops not tradeable when closed and more

KillerOfPie committed:
Chest names are being written, many bugs cleared, lots of todos, new ShopChest file, deleted ShopManager, lots of other changes

SparklingComet committed:
Refactor Executor.java

SparklingComet committed:
Add missing Shop#saveShop and #updateSign calls

SparklingComet committed:
Add tradeshop.edit perm & remove redundant StringBuilder#append calls

SparklingComet committed:
Update help message / fix colour coding

SparklingComet committed:
remove redundant dependency injection and use the '==' operator for enum comparison

KillerOfPie committed:
Removed old shop checks

KillerOfPie committed:
Fix for 'open' command not working, also sends a different message if it cant open shop

KillerOfPie committed:
Chests should reset names - untested

KillerOfPie committed:
Lots of smaller bug fixes, added limit to amount of shops that can be made with a single chest, fixed name reset when a sign or chest is destroyed, fixed shop not getting removed from file when destroyed

KillerOfPie committed:
Fixed BiTrades not doing reversed trade properly, added type 'switch' command and supporting methods

KillerOfPie committed:
Added ts 'what' command(opens inventory with items for trade), changed from storing items in map to base64 strings so that they can be properly loaded

SparklingComet committed:
Change from item1/item2 to product/cost

SparklingComet committed:
Change name of other half of dchest

SparklingComet committed:
Change name of other half of dchest

KillerOfPie commented on src/main/java/org/shanerx/tradeshop/objects/ShopChest.java:
```diff
@@ -134,6 +134,9 @@ public void resetName() {
 		if (bs instanceof InventoryHolder && bs instanceof Nameable) {
 			((Nameable) bs).setCustomName(((Nameable) chest.getState()).getCustomName().split(sep1)[0]);
+			((Nameable) Utils.getOtherHalfOfDoubleChest(chest).getState()).setCustomName(
+					((Nameable) Utils.getOtherHalfOfDoubleChest(chest).getState()).getCustomName().split(sep1)[0]);
```
this may cause an NPE if there isnt a double chest

KillerOfPie commented on src/main/java/org/shanerx/tradeshop/objects/ShopChest.java:
```diff
@@ -142,6 +145,7 @@ public void setName() {
 		BlockState bs = chest.getState();
 		if (bs instanceof InventoryHolder && bs instanceof Nameable) {
 			((Nameable) bs).setCustomName(getName());
+			((Nameable) Utils.getOtherHalfOfDoubleChest(chest).getState()).setCustomName(getName());
```
same here

KillerOfPie commented on src/main/java/org/shanerx/tradeshop/commands/Executor.java:
```diff
@@ -544,7 +546,7 @@ public boolean onCommand(CommandSender sender, Command cmd, String label, String
 				return true;
 
-			} else if (args[0].equalsIgnoreCase("getCI")) {
+			} else if (args[0].equalsIgnoreCase("getCI") || args[0].equalsIgnoreCase("items")) {
```
This command should get a single custom item not a list of items, also i think were going to retire custom items for now

SparklingComet committed:
Merge branch '2.0' of http://github.com/SparklingComet/TradeShop into 2.0

KillerOfPie committed:
Fully commented Shop file, changed vague or incorrectly name methods, added set/get amount methods for cost/product

KillerOfPie committed:
bug fixes, added protection for placing a chest next to existing shop(creating a double chest without being owner)

KillerOfPie committed:
Fixed bug wth items that have metadata but no custom name

KillerOfPie committed:
Fully implemented BukkitVersion obj for version checking
+ Code Cleanup

KillerOfPie committed:
Removed incorrect version check

KillerOfPie committed:
Fixed Version Check
Updated Maven to timestamp builds so that its easier to find the newest while testing

KillerOfPie committed:
Fixed errors during trade and other minor fixes
+ Added blank item to 'what' command
+ Changed lore for 'what' items
+ Updated Material names that changed in 1.13
+ Added 'hasProduct' and 'hasCost' methods to Shop
+ Added toString to 'getType' pulls for shop signs
+ Added below 1.13 check
+ Updated plugin.yml for 1.13

KillerOfPie committed:
Fixed long names not printing on sign

KillerOfPie committed:
New Objects and files for cleaning up command processing

KillerOfPie committed:
Started changing how commands work, almost done
+ Added command runner
+ Most permission/player checks are in command caller

KillerOfPie committed:
Finished moving all command operation to CommandCaller and
+ Deleted Executor
+ Changed commands the were add/remove owner to manager
+ Changed remove role commands in shop to boolean so that messages can be sent on failure more easily
+ Replaced Executor with CommandCaller in Tradeshop

KillerOfPie committed:
Buzzed mans cleanup
+ Removed unnecessary checks
+ Moved ill placed #replace calls
+ Removed Commands object from CommandRunner as it was unused
+ Fixed sendMessage calling itself
+ Removed unnecessary returns

KillerOfPie committed:
Small fixes for incorrect logic and such
+ took out unnecessary else
+ fixed logic for arg amount checking(it was reversed, whoops)
+ Made #checkPerm look for NONE perm as it wasnt clearing a perm check
+ #addProduct & #addCost now check for arguments before checking for int/mat

KillerOfPie committed:
Removed Potions file as it is no longer needed
+ Also removed only reference in Utils(didn't check if that effects anything but it shouldn't)

KillerOfPie committed:
Fix for long material names not showing up on signs
+ Fixed regex so it should be working again
+ Added setProduct & setCost aliases

KillerOfPie committed:
Lots of comments added for javadoc stuff, hopefully i did it all right

KillerOfPie committed:
Fix for long material names not showing up on signs(For real this time)
+ Set max length of string being set to 16 characters
+ Also learned the new lambda if method and used it

KillerOfPie committed:
Fixed BiTrade not working properly

KillerOfPie committed:
Fixed iTrade shops not being created without a chest
+ ShopType holds Permissions object instead of using it to store a Permission object

KillerOfPie committed:
Fixed error with breaking signs and chests
+ Cleaned out repetition in AdminEventListener
+ Added in getUUID() method when checking UUID(I swear im stupid sometimes)
+ Added check in ShopChest#resetName if chest doesn't have a custom name
+ Added check in ShopChest#resetName if chest doesnt have our separation char in its custom name

KillerOfPie committed:
Re-added support for entering items on initial sign creation
+ Added Shop#updateSign method that takes SCEvent specifically for updating sign at end of creation
+ Fixed incorrect comment in utils

KillerOfPie committed:
Fixed errors when making a sign with incorrect text on line #1/2
+ Removed check for owner/manager on what command
+ Added ShopCreateListener#lineCheck to see if lines have valid information to make an item
+ Shop#signUpdate sets lines to blank if the shop has no items for that line

KillerOfPie committed:
Added usable debug method to util that can be enabled in config
+ Added 'ENABLE_DEBUG' to settings, default to false
+ Trimmed fat from Utils
+ Added setting check to Utils#debug

KillerOfPie committed:
Update Copyright
+ Did it to every file this time not just LICENSE
+ Updated README.md

KillerOfPie committed:
Update Copyright
+ I put myself as an author in Idea finally, yaaaa

KillerOfPie committed:
Merge remote-tracking branch 'origin/2.0' into 2.0

KillerOfPie committed:
Fixed shop commands not checking players role
+ Fixed being able to add self as member/manger to any shop
+ Fixed being able to change product of any shop
+ Changed NOT_OWNER message to NO_SHOP_PERMISSION message
+ Changed CommandRunner#who catch from NoSuchFieldException to Exception aswho was through exceptions at random spots and it's not important enough for me to figure out now, the command works on signs and sometimes chests

SparklingComet commented on LICENSE:
```diff
@@ -186,7 +186,7 @@
       identification within third-party archives.
 
-   Copyright {yyyy} {name of copyright owner}
+   Copyright 2016-2019 SparklingComet @ http://shanerx.org & KillerOfPie @ https://killerofpie.github.io/
```
Good catch. Need to update that link though as I took it down.

SparklingComet replied:
Oh, also we need to add that to the single source files. I'll have my IDE automatically substitute the headers of  every file.

SparklingComet commented on README.md:
```diff
@@ -15,10 +15,10 @@ You may download the latest version from the [plugin's resource page on SpigotMC
+A full list of commands can be found [here](https://github.com/ShanerX/TradeShop/wiki/Commands), and Permissions list can be found [here](https://github.com/ShanerX/TradeShop/wiki/Permissions).
 
-#### The current version is marked as BETA as significant changes/fixes were introduced. This means that it is very likely that there will be bugs.
+#### The current version is marked as DEV as significant changes/fixes were introduced. This means that it is very likely that there will be bugs.
```
Let’s just not forget changing it back to STABLE or BETA before releasing. No official releases should ever be marked as DEV

KillerOfPie replied:
made todo

SparklingComet commented on pom.xml:
```diff
@@ -6,16 +6,18 @@
     <groupId>org.shanerx</groupId>
     <artifactId>tradeshop</artifactId>
-    <version>1.8.1</version>
+    <version>2.0.0-DEV</version>
```
same here.

KillerOfPie replied:
made todo

SparklingComet commented on src/main/java/org/shanerx/tradeshop/TradeShop.java:
```diff
@@ -22,404 +22,82 @@
+		pm.registerEvents(new AdminEventListener(this), this);
+		pm.registerEvents(new ShopCreateListener(), this);
+		pm.registerEvents(new ShopTradeListener(), this);
+		pm.registerEvents(new CustomInventoryListener(), this);
```
This change was long due, good one

KillerOfPie replied:
thanks

SparklingComet commented on src/main/java/org/shanerx/tradeshop/commands/CommandCaller.java:
```diff
@@ -0,0 +1,160 @@
+            case REMOVEMEMBER:
+                cmdRnnr.removeMember();
+                break;
+
```
Much cleaner, +1

KillerOfPie replied:
thanks

SparklingComet commented on src/main/java/org/shanerx/tradeshop/commands/CommandCaller.java:
```diff
@@ -0,0 +1,160 @@
+    public boolean checkPerm() {
+        if (!cmdPass.getSender().hasPermission(command.getPerm().getPerm()) && !command.getPerm().equals(Permissions.NONE)) {
+            cmdPass.getSender().sendMessage(Message.NO_COMMAND_PERMISSION.getPrefixed());
+            return false;
```
I am slightly confused as how this permission check works. Weren't there different permission nodes for different subcommands? I don't understand where/when the checks are performed.

Edit: some are performedin through the `CommandRunner` instance (help command), but not all.

KillerOfPie replied:
The main permission to run any command is checked by that before even creating the CommandRunner object to lessen unnecessary object creation. CommanRunner#help checks perms only for sending help for the commands that the player has access to. That is the only method in CommandRunner that checks perms.

SparklingComet commented on src/main/java/org/shanerx/tradeshop/commands/CommandRunner.java:
```diff
@@ -0,0 +1,765 @@
+    }
+
+
+}
```
Alright, all of this make sense. I did not test it nor check the logic of every statement as you reported it to be working perfectly.

While organizing everything this way does make sense, all these subcommands implemented in the same file does lead to a relatively large source file. I am aware that in the past we used to split things up, which lead to a messy codebase and large amounts of time wasted wondering whether something had been implemented successfully, and if so, where.

Nevertheless, such a large file can make maintenance difficult. Either way, the implementation is clean and the code pretty enough so for now we can keep it this way I guess.

KillerOfPie replied:
The reason i ended up with this was because I didn't want to have 12 different objects that would only be used to run a command and had no need to return anything as it would cause unnecessary overhead for the system. Also decided that while this is similar to how it was it is much easier to find specific sections and follow the code as its not all in nested if statements that just meld together.

SparklingComet commented on src/main/java/org/shanerx/tradeshop/enumys/Commands.java:
```diff
@@ -0,0 +1,164 @@
+    REMOVECUSTOMITEM(Lists.newArrayList("removeCustomItem", "removeItem", "delItem"), Permissions.ADMIN, 2, 2, false),
+    GETCUSTOMITEMS(Lists.newArrayList("getCustomItems", "getCIs"), Permissions.ADMIN, 2, 2, true),
+    RELOAD(Lists.newArrayList("reload"), Permissions.ADMIN, 1, 1, false),
+    SWITCH(Lists.newArrayList("switch"), Permissions.EDIT, 1, 1, true);
```
I like this.

KillerOfPie replied:
thanks

SparklingComet commented on src/main/java/org/shanerx/tradeshop/enumys/Commands.java:
```diff
@@ -0,0 +1,164 @@
+     * @return true if command requires player
+     */
+    public boolean needsPlayer() {
+        return needsPlayer;
```
Smart system, +1

KillerOfPie replied:
thanks

SparklingComet commented on src/main/java/org/shanerx/tradeshop/enumys/Message.java:
```diff
@@ -0,0 +1,183 @@
+	public static String colour(String x) {
+		return ChatColor.translateAlternateColorCodes(COLOUR_CHAR, x);
+	}
+}
```
This too was long overdue, good job removing stuff from the main class.

KillerOfPie replied:
thanks

SparklingComet commented on src/main/java/org/shanerx/tradeshop/enumys/ShopRole.java:
```diff
@@ -0,0 +1,58 @@
+@SuppressWarnings("unused")
+public enum ShopRole {
+
+	OWNER(true, true, true, false),
```
Any chance we can get a comment up there with the name of every argument? I am aware we have the constructor following within 10 lines of code, but this sequence of booleans does look weird to the eye.

Not at all a requirement, though.

KillerOfPie replied:
I commented it, maybe not the best and we can change it later if it doesn't help. I didn't notice how daunting it look before because Idea labels each one.

SparklingComet commented on src/main/java/org/shanerx/tradeshop/enumys/ShopStatus.java:
```diff
@@ -0,0 +1,50 @@
+
+public enum ShopStatus {
+	OPEN("&a"),
+
```
Sorry for the useless feedback on this one, but the empty line does look particularly painful to me. Do what you will, though.

KillerOfPie replied:
done

SparklingComet commented on src/main/java/org/shanerx/tradeshop/listeners/AdminEventListener.java:
```diff
@@ -0,0 +1,160 @@
+			event.setCancelled(true);
+			player.sendMessage(Message.NO_TS_DESTROY.getPrefixed());
+		}
+	}
```
What exactly happens in the case of Double Chests? Where exactly are the necessary checks performed? ie. looking for the other half to make sure it is not a shop etc.

KillerOfPie replied:
The answer was utils, but I just moved them due to there being a wrapper in ShopChest and them making more sense there.

SparklingComet commented on src/main/java/org/shanerx/tradeshop/listeners/CustomInventoryListener.java:
```diff
@@ -0,0 +1,42 @@
+
+		Inventory inventory = event.getInventory();
+
+		if (ChatColor.stripColor(inventory.getName()).endsWith("Shop                                 ")) {
```
Beg pardon, but what is this?

KillerOfPie replied:
Prevents people from taking things out of our custom inventories, kinda necessary otherwise you could literally just pull the items we put in them out. Also there is a counted amount of spaces after the name so i know its our inventory, don't know how many it is anymore though. But that is the only way to distinguish between ours and others without having a terrible looking filled title on it.

SparklingComet replied:
Alright, makes sense. Just got confused.
Would be more practical to have it defined as a constant somewhere, since I assume it appears elsewhere, where the title is set. Would make changing the identifier easier.

SparklingComet commented on src/main/java/org/shanerx/tradeshop/listeners/ShopCreateListener.java:
```diff
@@ -0,0 +1,148 @@
+
+		return product;
+	}
+}
```
Has anything changed here? I can't figure out what's new about this, in terms of functionality.

KillerOfPie replied:
First it got combined and stripped, then i added back the ability to type the items directly on the sign. Which is what lineCheck is doing. however if line[1] and line[2] are not typed correctly to be used they will be wiped and the shop will create without items.

SparklingComet replied:
What exactly do you mean with "the shop will create without items"?

KillerOfPie replied:
The shop will create with line 1 & 2 empty and the status closed(which it will do whether or not items are present) and the player would have to use /addcost and /addproduct to set the items for the shop before opening it.

SparklingComet commented on src/main/java/org/shanerx/tradeshop/listeners/ShopTradeListener.java:
```diff
@@ -0,0 +1,411 @@
+                    .replace("{SELLER}", shop.getOwner().getPlayer().getName()));
+		}
+	}
+}
```
same as above ^^

KillerOfPie replied:
For the most part no, just all new when the files got merged

SparklingComet commented on src/main/java/org/shanerx/tradeshop/objects/CustomItemManager.java:
```diff
@@ -0,0 +1,179 @@
+
+	public String colour(String x) {
+		return ChatColor.translateAlternateColorCodes(COLOUR_CHAR, x);
+	}
```
Hadn't we already other String colour shortcuts lying around elsewhere? No harm done, but no point making this `public`

KillerOfPie replied:
it wasn't even being used, its deleted now.

SparklingComet commented on src/main/java/org/shanerx/tradeshop/objects/ListManager.java:
```diff
@@ -0,0 +1,126 @@
+						Material.BROWN_SHULKER_BOX,
+						Material.GREEN_SHULKER_BOX,
+						Material.RED_SHULKER_BOX,
+						Material.BLACK_SHULKER_BOX));
```
There are better ways to do this, but fine. We probably discussed this after the release of 1.9.

KillerOfPie replied:
I'm doing it this way so that each shulker type doesn't needed to be typed into the config. Although i guess making an array for that is useless and i should just call #add for each one. *changed

SparklingComet replied:
I am aware of that, but still there are better ways of determining whether the block is some kind of shulker box which do not involve a dozen enum comparisons.

KillerOfPie replied:
So all this piece is doing is adding the shulker materials to the list that also contains all other valid storage types for the shops. and checking materials is the easiest way since im not just checking for shulkers.

SparklingComet commented on src/main/java/org/shanerx/tradeshop/objects/Shop.java:
```diff
@@ -0,0 +1,756 @@
+		updateSign();
+		saveShop();
+	}
+}
```
Were any mechanics changed or was this just cleaned up/moved to another package?

KillerOfPie replied:
Besides being made for 2.0 I think the only changes recently are minor fixes.

SparklingComet commented on src/main/java/org/shanerx/tradeshop/objects/ShopChest.java:
```diff
@@ -0,0 +1,184 @@
+	private Location loc;
+	private Block chest;
+	private UUID owner;
+	private String sep1 = "\\$ \\^", sep2 = ":";
```
Beg pardon, but what are these? I understand they are separators for data meant to be held in a single string, but perhaps it would be better to add a comment or renaming the variables - make them constants, please - to clarify what they keep apart (names? uuids? coordinates?)

KillerOfPie replied:
Renamed sectionSeparator and titleSeparator, as thats what they are separating

SparklingComet replied:
Alright, better still if made `final`. Not of utmost importance, though.

SparklingComet commented on src/main/java/org/shanerx/tradeshop/objects/ShopChest.java:
```diff
@@ -0,0 +1,184 @@
+		return null;
+	}
+
+	public Block getOtherHalfOfDoubleChest() {
```
This answers a question I asked earlier, nevermind.

KillerOfPie replied:
Moved the actual methods here instead of this just wrapping the Utils methods

SparklingComet commented on src/main/java/org/shanerx/tradeshop/objects/ShopLocation.java:
```diff
@@ -0,0 +1,97 @@
+
+	public static ShopLocation deserialize(String loc) {
+		if (loc.startsWith("l")) {
+			String[] locA = loc.split("_"); //Keep same as div
```
then why not put `div` in it? instead of keeping the literal.

KillerOfPie replied:
Because that method needs to be static as it needs to be run before you have an object, also div doesn't exist in that file anymore so i just removed the comment

KillerOfPie replied:
I lied, i just put dive back in but either way deserialize returns the object we will need from the string. I may make a constructor that takes the string later but i prefer this as i can return and check for null to check if the string was a location/chunk or not

SparklingComet commented on src/main/java/org/shanerx/tradeshop/objects/ShopChunk.java:
```diff
@@ -0,0 +1,82 @@
+
+	private World world;
+	private int x, z;
+	private String div = "_";
```
* please make it constant
* please remove the `private` modifier to allow (default) package access, since it is supposed to be used in `ShopLocation`.

KillerOfPie replied:
same as above, and got changed to match location as well since deserialize is useless to the object after creation

SparklingComet commented on src/main/java/org/shanerx/tradeshop/objects/ShopLocation.java:
```diff
@@ -0,0 +1,97 @@
+	}
+
+	public String serialize() {
+		return "l" + "_" + world.getName() + "_" + x + "_" + y + "_" + z;
```
Please use `div` here too. Using the literal instead of the previously defined constant/variable is seldom an improvement.

KillerOfPie replied:
done, took it out at some point. idk why tho

SparklingComet commented on src/main/java/org/shanerx/tradeshop/objects/ShopUser.java:
```diff
@@ -0,0 +1,82 @@
+	}
+
+	private void fix() {
+		if (player == null && playerUUID != null && playerUUID != "") {
```
I could be wrong, but I cannot remember why the last check would be necessary

KillerOfPie replied:
first off changed to #equalsIgnoreCase, and secondly it's because im paranoid and also Idea doesn't tell me its unnecessary so i assume there is a way for it to happen.

SparklingComet commented on src/main/java/org/shanerx/tradeshop/utils/BukkitVersion.java:
```diff
@@ -0,0 +1,152 @@
+
+		return map;
+	}
+}
```
Do we need this? After all, support for anything below 1.13 is being dropped.
And if we are thinking about the future, we can always add it back. After all, it seems you have removed all calls to this class and its members.

KillerOfPie replied:
It is being used. just not fully, #isBelow is what is checking our versions in the main file and if we need to check version for features later this gives us the ability

SparklingComet commented on src/main/java/org/shanerx/tradeshop/utils/ItemSerializer.java:
```diff
@@ -0,0 +1,89 @@
+	 * Original code from https://gist.github.com/graywolf336/8153678
+	 * Tweaked for use with single itemstacks
+	 *
+	 */
```
For very tedious reasons, we are technically required to put this above the class name - it was probably my mistake, as I added this.

KillerOfPie replied:
I understand why

SparklingComet commented on src/main/java/org/shanerx/tradeshop/utils/JsonConfiguration.java:
```diff
@@ -0,0 +1,121 @@
+		shop.fixAfterLoad();
+		return shop;
+	}
+}
```
Is any of this new? I seem to recall we had this before, and yet git marks it as new. Though perhaps I am wrong, as I am using something like this in other plugins too.

KillerOfPie replied:
It's new to 2.0 and you're looking at all the changes for 2.0

SparklingComet commented on src/main/java/org/shanerx/tradeshop/utils/Updater.java:
```diff
@@ -0,0 +1,190 @@
+
+		BEHIND,
+
+		UNKNOWN
```
those empty lines are painful to look at, but not the point.
Is there anything new here? this file was moved, and the previous one had fewer lines.

KillerOfPie replied:
Moved i think is all, also you put those in there :P
Fixed now tho.

SparklingComet replied:
I am aware of the inconsistency in my own code in the past. Still would like to figure out what was added to the file.

SparklingComet commented on src/main/java/org/shanerx/tradeshop/utils/Utils.java:
```diff
@@ -0,0 +1,422 @@
+		}
+
+		return false;
+
```
pointless whitespace

KillerOfPie replied:
done

SparklingComet commented on src/main/java/org/shanerx/tradeshop/utils/Utils.java:
```diff
@@ -0,0 +1,422 @@
+			plugin.getLogger().log(Level.WARNING, text);
+		}
+	}
+}
```
Since we no longer have a massive file with every utility function thrown into it, and since it is inside the `utils` package, perhaps this can be renamed into something more meaningful.
Or we could delete the file and move the functions to other files, ie. `ShopChest`.

KillerOfPie replied:
that's the idea, just wasn't the biggest priority once it was trimmed down this far

SparklingComet replied:
Fair enough

SparklingComet commented on src/main/resources/LICENSE:
```diff
@@ -186,7 +186,7 @@
       identification within third-party archives.
 
-   Copyright {yyyy} {name of copyright owner}
+   Copyright 2016-2019 SparklingComet @ http://shanerx.org & KillerOfPie @ https://killerofpie.github.io/
```
I am changing the url before the release

KillerOfPie replied:
cool

SparklingComet commented on src/main/resources/plugin.yml:
```diff
@@ -1,9 +1,10 @@
 name: TradeShop
-version: 1.8.1
+version: 2.0.0-DEV
```
same here, ought to remove the `DEV` tag before releasing

KillerOfPie replied:
added todos in TradeShop so that we wont forget, hopefully

SparklingComet commented on src/main/resources/plugin.yml:
```diff
@@ -1,9 +1,10 @@
+version: 2.0.0-DEV
 main: org.shanerx.tradeshop.TradeShop
+api-version: 1.13
+load: POSTWORLD
```
I am not 100% sure we need this, but okay.

KillerOfPie replied:
the load we more then likely don't need but we don't need to load pre world so i would prefer we load after it, just in case. The API version we do need, Spigot doesn't like it otherwise since we are using 1.13 material names and not the legacy ones

SparklingComet replied:
Okay, was only referring to the load clause.

SparklingComet reviewed (approved):
I have a few questions, probably due to not contributing anything for a long time.
This pull request features massive improvements, excellent work.
Good job on cleaning up the mess, too.

SparklingComet commented:
Also, if you are going to merge, please do not squash. You probably weren't going to, but just to be sure.

KillerOfPie commented:
I hate squashing :P also still may be some more changes before completely finalizing it would rather not have a beta tag on this version too

KillerOfPie committed:
Fixes from review
+ Added comment for enums in ShopRole
+ Fixed spacing in ShopStatus
+ removed CustomItemManager#colour as it was unused and unnecessary
+ Changed ListManager addAll call to add call for each item as making a list was unnecessary
+ Renamed separators in ShopChest
+ Moved double chest methods from utils to ShopChest
+ Made div final
+ Made deserialize static as it should only be used without an object
+ Re-Added div to ShopLocation
+ Fixed logic in ShopUser
+ Removed whitespace from Updater
+ Removed double chest methods from Utils
+ Added todo to TradeShop to change Dev tag before finish

SparklingComet committed:
Update Copyright headers to include KillerOfPie

KillerOfPie committed:
Should fix thecompile errors after moving the methods from Utils
+ Not really what i want will probably change once i can think of a way to achieve this without Utils and static

KillerOfPie committed:
Possible fix for NPE on SU:56 -> CR:262

KillerOfPie committed:
Fix for NPE on SU:56 -> CR:262

KillerOfPie committed:
Adds support for 1.14 Sign types while still supporting 1.13
+ Added ShopSign object
+ Added Signs enum within ShopSign
+ Signs holds the new and old sign types and their min/max versions
+ ShopSign checks what types it should be using based on version
+ ShopSign#getSignTypes returns List<Material> of all applicable sign types

KillerOfPie committed:
Fixes support for 1.14 Sign types while still supporting 1.13
+ Added some debug, may/may not remove later as it requires the debug setting to be on
+ Changed logic in a way that doesn't really matter
+ Initiated minVer/maxVer because I forgot to
+ Made ^ fixed arrays to be more efficient
+ Fixed broken regex '.' != '[.]'

KillerOfPie committed:
Added Tab auto-complete for all sub commands
+ Added CommandTabCaller
+ Added CommandTabComplete
+ Added Shop#getManagersNames & Shop#getMembersNames to return list of names
+ Removed a blank line from CommandCaller

KillerOfPie committed:
Minor changes to tab completion
+ Tab complete for sub command names, gets the first name in list instead of enum name(so when we make commands customizable it auto-completes the custom name)
+ Added Commands#getFirstName to retrieve the first command name in the list

KillerOfPie committed:
Moved blacklist checking methods
+ Added AIR, VOID_AIR, CAVE_AIR to default blacklist
+ Changed BlackList from String to Material
+ Moved blacklist checks from Utils to ListManager

KillerOfPie committed:
Removed CustomItems
+ Removed all custom item related code that I could find

SparklingComet committed:
Implement backbone of the addon custom subcommand framework.

SparklingComet committed:
Have the addon framework for subcmds hook into the usage and permission fields
to provide smarter messages, as well as extend the help message to include custom subcommands

SparklingComet committed:
Fix major Iterator<?> malfunction and add AddonManager and PlayerTradeEvent
this lays the groundwork for the addon management infrastructure and the custom events fired by the plugin.

KillerOfPie commented on src/main/java/org/shanerx/tradeshop/commands/CommandCaller.java:
```diff
@@ -53,14 +53,23 @@ public boolean onCommand(CommandSender sender, Command cmd, String label, String
+	        CustomCommandHandler handler = CustomCommandHandler.getInstance();
+	        String subcmd = cmdPass.getArgAt(0);
+
+	        if (!handler.isAvailable(subcmd)) {
```
Either isAvailable is returning backwards or this shouldn't be !

SparklingComet replied:
isAvailable returns true if that subcommand has not been registered. Hence, if the value is false, it means the command has already been registered by some addon or by the plugin itself (aka it is one of the entries of the enum Commands). Clearly, the second option is not possible as if it were, the null check would fail. 

The conclusion is that if isAvailable returns false, the subcommand exists.

isAvailable refers to whether or not an addon may register a certain subcommand, not whether that subcommand is available for execution.

This will become clear once I write the docs.

KillerOfPie replied:
ah so is available to register, got it, makes more sense now

KillerOfPie commented on src/main/java/org/shanerx/tradeshop/commands/CommandCaller.java:
```diff
@@ -53,14 +53,23 @@ public boolean onCommand(CommandSender sender, Command cmd, String label, String
+	        String subcmd = cmdPass.getArgAt(0);
+
+	        if (!handler.isAvailable(subcmd)) {
+		        handler.getExecutable(subcmd).run(args);
```
I havnt looked at #run but can we make that return a boolean and then just return that that way subCommands would be made the same as a regular command and if the maker returns false it just gives the default message.

SparklingComet replied:
That is precisely how #run is intended to work.

KillerOfPie replied:
but then wouldn't you just be returning #run?

KillerOfPie replied:
is #run running the command or registering the command?

SparklingComet replied:
Running it.
This is weird, I should have had `if (!...run(...)) { /* send msg for invalid args */ }`
I was actually sure I did it like that.... will fix that tomorrow.

KillerOfPie replied:
or you could just have it return #run and then the return is wholly on the run method

SparklingComet replied:
True, but I would like every command that starts with /ts to have the same “invalid args” message. Having the addon do it leads to inconsistent behaviour, which can be baffling to the common user.

If, however, the developer of the addon truly wishes to handle usage messages they simply have to return true and they are free to output whatever they desire.
After all, the return value is not really used for anything, apart from this

KillerOfPie replied:
then we can catch the #run return and if false send our invalid message. I think this is the same thing you were going to do, so that sounds good.

KillerOfPie commented on src/main/java/org/shanerx/tradeshop/framework/CustomCommandHandler.java:
```diff
@@ -0,0 +1,78 @@
+
+	public void registerCommand(TradeCommand subCmd) {
+		if (!isAvailable(subCmd.getCmd())) {
+			throw new IllegalArgumentException("Another addon has already register a subcommand with this name already exists!");
```
Wording is a little weird on the message.

SparklingComet replied:
You are right, I got distracted while writing it.

KillerOfPie commented on src/main/java/org/shanerx/tradeshop/commands/CommandCaller.java:
```diff
@@ -154,4 +160,18 @@ public boolean checkPerm() {
+     * @return true if permission is NONE or sender has permission
+     */
+    public boolean checkPerm(TradeCommand command) {
+        if (!cmdPass.getSender().hasPermission(command.getPermission()) && !command.getPermission().equals(Permissions.NONE)) {
```
Can we make this check if the add-on adds a permission for the command as well?

SparklingComet replied:
What do you mean? command.getPermission() is the permission for the subcommand, provided by the addon.

KillerOfPie replied:
but this method is checking against the command enum is it not? If it is, it isnt actually checking for the add-on sub commands

SparklingComet replied:
I don’t understand what you mean. The function TradeCommand#getPermission() is most certainly meant to return the permission for that command.

KillerOfPie replied:
I see now, didnt register that its using TradeCommand not Commands. can we change the variable to tradeCommand instead of command so that its a little easier to read?

SparklingComet replied:
Yes, now there are two `checkPerms` methods, distibguished by their argument types.

KillerOfPie replied:
yeh, i think changing the variable names will make it easier to tell the difference between the 2 methods when glancing at them

SparklingComet replied:
Ok, will do.

KillerOfPie commented on src/main/java/org/shanerx/tradeshop/commands/CommandRunner.java:
```diff
@@ -104,7 +106,19 @@ public void help() {
 
-        sendMessage(sb.toString());
+	    CustomCommandHandler handler = CustomCommandHandler.getInstance();
+        while (handler.iter().hasNext()) {
```
Is this a per command help message?

SparklingComet replied:
Yes. The addon provides the help message (description + usage) and I am simply including it in the output for /ts help

KillerOfPie replied:
If we could do that for our own sub commands to that would be nice. I was already thinking about doing it so that people could use '/ts help addProduct' to get a cleaner description and usage and the reg help command could just list commands.

SparklingComet replied:
I thought about that, but I did not see a reason to do it.
Firstly, our own, “official” commands are displayed exactly the same way; secondly, for users who are trying to figure out how to use a command, this seems to be the easiest way to relay the information.
Is having yet _another_ command with _another_ help message really worth the hassle?
I think there is a chance that it would complicate things for end users instead of making them simpler.

SparklingComet replied:
Wdym? I am only extending the present help to include addon commands.
A per command help would most certainly be an elegant solution, but would it be necessary? What information apart from the description, the usage/args and the name of the potential addon that introduces it (all of which are already displayed) could we display in the per command help?

KillerOfPie replied:
Its not necessarily more information as much as a cleaner look, I really hate how our current help menu looks, it just feels too full. Also most of the time you use a help command its because you're looking for usage on a single command or you just cant remember what a command is called. so having a straight list for 'help' and usage on 'help <command>' would be better.

SparklingComet replied:
Fair point.

So you are suggesting we only have the command and the description in help, and the usage explained in the specific help for the command? That could indeed be cleaner, hadn‘t thought about that.

KillerOfPie replied:
I was actually going more with just command, but command and description could be fine too

SparklingComet replied:
I will take care of it tomorrow.

SparklingComet committed:
Comply with review & remake the help command

SparklingComet committed:
Fix copyright headers and update the year

SparklingComet committed:
Fix missing call for usage()

KillerOfPie commented on src/main/java/org/shanerx/tradeshop/enumys/Commands.java:
```diff
@@ -34,49 +34,65 @@
-    WHAT(Lists.newArrayList("what", "peek", "windowShop", "shop", "view"), Permissions.INFO, 1, 1, true),
-    RELOAD(Lists.newArrayList("reload"), Permissions.ADMIN, 1, 1, false),
-    SWITCH(Lists.newArrayList("switch"), Permissions.EDIT, 1, 1, true);
+    HELP(Lists.newArrayList("help", "?"), Permissions.HELP, 1, 2, false, "Display help message", "/ts help [command]"),
```
Can we pick either /ts or /tradeshop for these
Also we should put in a placeholder for the command name and just have it replace when it is retrieved that way we can use all the aliases  like you have for the later commands. This also leads to having support for servers changing the commands(for language or appearance purposes)
Last some of the commands are missing the /, i think it looks fine either way but we do need to pick a way

I can fix all these if you want just let me know

SparklingComet replied:
You are right about having to be consistent, my mistake. I think displaying /tradeshop is better than its short version /ts.

I don‘t think placeholders are necessary, it would be useful for non English servers, but it would force us to bloat the source code by complicating stuff.

KillerOfPie replied:
the only difference would be a fairly simple replace in #usage and a small loop for building the string. I was planning on making the command names customizable by the server anyways, so this would just be a step that's probably going to be taken anyways.

and using the full name for command usage is usually what i default to

SparklingComet replied:
We could do that. However, since that isn’t a priority I would prefer to add it in 2.1.0.

It would be a significant feature (despite its simple implementation) and we are already introducing major changes.

No point rushing it in now.

KillerOfPie replied:
kk im down with that

KillerOfPie replied:
I wasn't planning on fully implementing it yet, just was saying since we're making this now we might as well make it in a way that doesn't require use to change it right after we release 2.0

KillerOfPie commented on src/main/java/org/shanerx/tradeshop/commands/CommandRunner.java:
```diff
@@ -77,52 +77,43 @@ public void help() {
+	    }
+	    else if (handler.isNativeCommand(subcmd)) {
+		    Commands cmd = Commands.getType(subcmd);
+		    sendMessage(Message.colour(String.format("&6Showing help for &c%s&r\n&b%s &e - %s", subcmd, cmd.getUsage(), cmd.getDescription())));
```
So does format replace %s in the same order you give it strings?

SparklingComet replied:
Indeed

KillerOfPie commented:
Congrats on commit #100, I wanted it but good job.

KillerOfPie committed:
Changed up per command help
+ Added Commands#getALiases
+ Commands#getUsage now replaces $cmd$ placeholder with the first command name

KillerOfPie committed:
Fix for commands not working on wall signs
+ added WALL_SIGN to signTypes
+ Also moore debug lines that i may or may not remove later

KillerOfPie committed:
Fix for commands not working will looking at shop chest
+ Reformatted file, should really do that separatly so the it doesnt show every line as a change -_-
+ Added findShop method to CommandRunner so that every command doesnt do the same 12 lines of code
+ Fixed Shop#loadShop endlessly calling each other

KillerOfPie committed:
Fixed bugs in #addProduct/#addCost
+ Fixed permission logic so the #addProduct actually runs
+ Fixed bug in #addProduct/#addCost the allowed a player to change the amount of items in the held stack

KillerOfPie committed:
Made Command tabCompletion complete based on what's been input
+ Moved Shop#saveShop & Shop#ipdateSign into Shop#set methods
+ Added Commands#isPartialName to check all names/aliases for a specified string

KillerOfPie committed:
Changed #removeManager & #removeMember to #removeUser
+ Changed Shop storage of users to UUIDs from ShopUsers
+ Fixed tab completion of online players for #addMember and #addManager
+ Fixed tab completion of users on a shop for #removeUser
+

KillerOfPie committed:
Fixed help message and tab complete for help message
+ Moved permission checking to Commands#checkPermission

KillerOfPie committed:
Minor Tweaks and fixes
+ Updated setup message
+ Added prefix to bug command

KillerOfPie committed:
Fixed reload command not reloading messages or lists from settings

SparklingComet committed:
Merge branch 'master' into 2.0

KillerOfPie committed:
Update README.md

SparklingComet committed:
Unhook disabled plugins and
stop unhooked addons from registering commands

SparklingComet committed:
Merge branch '2.0' of https://github.com/SparklingComet/TradeShop into 2.0

KillerOfPie committed:
added line breaks
